applications/nrf_desktop/src/modules/hid_forward.c

```diff
@@ -1187,81 +1228,87 @@ static void hogp_out_rep_write_cb(struct bt_hogp *hogp, struct bt_hogp_rep_info
 
 			if (send_err) {
 				LOG_ERR("Cannot forward HID report (err: %d)", send_err);
 			}
 
-			sys_slist_remove(&per->enqueued_out_reports, prev_node, cur_node);
-			k_free(cur_report->data);
-			k_free(cur_report);
-			break;
+			WRITE_BIT(per->enqueued_out_reports_bm, sent_report_id, 0);
+			return;
 		}
-
-		prev_node = cur_node;
 	}
+	LOG_ERR("Enqueued report not found.");
+	WRITE_BIT(per->enqueued_out_reports_bm, sent_report_id, 0);
 }
 
-static void append_hid_out_report(sys_slist_t *list, const uint8_t *data, size_t size)
+static void send_empty_hid_out_reports(struct hids_peripheral *per, struct subscriber *sub)
 {
-	struct enqueued_out_report *item = k_malloc(sizeof(*item));
+	struct bt_hogp *per_hogp = &per->hogp;
 
-	if (!item) {
-		LOG_ERR("Failed to allocate memory to forward HID output report");
-		return;
-	}
+	for (size_t i = 0; i < ARRAY_SIZE(sub->out_reports); i++) {
+		if (sub->saved_out_reports_bm & BIT(i)) {
+			uint8_t report_id = sub->out_reports[i][0];
+			int size = get_output_report_size(report_id);
 
-	item->data_size = size;
-	item->data = k_malloc(size);
+			__ASSERT_NO_MSG(size >= 0);
+			uint8_t empty_data[size];
 
-	if (!item->data) {
-		k_free(item);
-		LOG_ERR("Failed to allocate memory to forward HID output report");
-		return;
-	}
+			memset(empty_data, 0, sizeof(empty_data));
+			empty_data[0] = report_id;
+			int err = send_hid_out_report(per_hogp, empty_data, (size_t)size);
 
-	memcpy(item->data, data, size);
-	sys_slist_append(list, &item->node);
-};
+			if (err) {
+				LOG_ERR("Failed to forward output report (err: %d)", err);
+			}
+		}
+	}
+}
 
-static void enqueue_hid_out_report(sys_slist_t *list, const uint8_t *data, size_t size)
+static void save_hid_out_report(struct subscriber *sub, const uint8_t *data, size_t size)
 {
 	uint8_t report_id = data[0];
-	bool updated = false;
+	__ASSERT_NO_MSG(get_output_report_size(report_id) >= 0);
+	__ASSERT_NO_MSG(size == (size_t)get_output_report_size(report_id));
 
 	/* If report is already present, update the data associated with report. */
-	sys_snode_t *cur_node;
-
-	SYS_SLIST_FOR_EACH_NODE(list, cur_node) {
-		struct enqueued_out_report *queued_rep = CONTAINER_OF(cur_node,
-								      struct enqueued_out_report,
-								      node);
-		uint8_t queued_report_id = queued_rep->data[0];
-
-		if (queued_report_id == report_id) {
-			__ASSERT_NO_MSG(size == queued_rep->data_size);
-			memcpy(queued_rep->data, data, size);
-			updated = true;
-			break;
+	for (size_t i = 0; i < ARRAY_SIZE(sub->out_reports); i++) {
+		if ((sub->saved_out_reports_bm & BIT(i)) && (report_id == sub->out_reports[i][0])) {
+			memcpy(sub->out_reports[i], data, size);
+			return;
 		}
 	}
 
-	if (!updated) {
-		append_hid_out_report(list, data, size);
+	for (size_t i = 0; i < ARRAY_SIZE(sub->out_reports); i++) {
+		if (!(sub->saved_out_reports_bm & BIT(i))) {
+			memcpy(sub->out_reports[i], data, size);
+			WRITE_BIT(sub->saved_out_reports_bm, i, 1);
+			return;
```

Comment:
Maybe we could take index in array based on position of report ID in output_reports? I think making it consistent would make the code more straightforward and it could prevent errors in the future.

Reply:
We allocate memory for every output report anyway